rfcs/013 - Datastore Revisited RFC.md

```diff
@@ -0,0 +1,137 @@
+
+In this case, the `namespace` function would take not only the name parameter, but the type info for the kind of variable to store in it, like `namespace<Array<int64>>('variousDigitsOfPi')`, then this type would internally be applied to the namespace through some syntax similar to the `type variableName` before so it can be stored in the datastore opcodes, but it would also be used to make sure writes and reads are sound, always providing values of the correct type.
+
+The biggest issue here is: it adds a new "color" of function, and there will be pressure to add something like an `fn foo<A as Orderable>(...` syntax so the generics defined for the function are also interface constrained so the function knows what to do with it besides pass it around or poke at it in some way or the other.
+
+It'll make the language much more complex, but will the utility gain be worth the complexity? It would make constructor functions much simpler, eg instead of `newHashMap(firstKey, firstVal)` we get `new<HashMap<KeyType, ValType>>()` without requiring the first key-val pair to be defined.
+
+### Compiler-level hackery
+
+We could do it with neither of the above via effectively hackery within the compiler; the `ds*` opcodes get a special behavior that always appends the type information to the `namespace` string no matter what and we just have some magic behavior that "just works" but doesn't trust the end user to have access to that power.
+
+I'm including this for completeness, but that is the "Go Way" and in my book "The Go Way Is Stupid."
+
+### Rust-style impl-ish functionality
+
+This is similar to the Generic Functions approach but more constrained. Rust functions within an `impl` block are bound to a specific type. If that function begins with a variable of the same type named `self`, then it can be used as a method off of a variable of that type, if not, it can only be used by invoking it from the type itself with a double-colon, like `HandlerMemory::string_to_hm(...`. While this reduces the regularity of Alan, we could have such type-bound functions and require that each type that wants to be stored in datastore has to re-implement the 5 datastore functions, likely by adding the appropriate metadata to the namespace string.
+
+This has the downside of requiring explicit boilerplate code for new types that want to be stored in datastore (and therefore can be explicitly written incorrectly and not actually do the right thing) and also breaking method syntax for these functions, but it doesn't require the magic of the Generic Functions or even the `type variableName` statement.
+
+This boilerplate could be covered up if/when we add macro functionality to the language, which would also take us further along the path towards Rust.
+
+### Namespace *is* Type
```

Comment:
not too ecstatic about this, since it reduces functionality for users (two datastores that use the same type for different reasons)